### ISSUE TITLE
[rllib] PPO performance degrades in Rllib, not in Stable baselines

### ISSUE THREAD

**Issue by fredrikbissmarckatfoise** [ray-project/ray#9951]
Hi,

I am just starting Rllib but know reinforcement learning well.

I have a problem with convergence in a custom task trained by PPO. In Stable baselines framework, while not always monotone rise in reward, the same agent converges at near optimal behavior (about 2M timesteps) . In Rllib, near optimal behavior (maximum reward) is also obtained, but then performance degrades quickly to lowest possible reward.

Is there an obvious reason for this? Or should I suspect it is entirely task specific? 

Are there major differences in implementation of PPO in stable baselines and in Rllib? (I got the same result with A2C too).
I use default parameters except for vf_clip_param which I set to a large number. The PPO parameters are comparable to
Stable baselines (PPO2). As for the policy, I need to look at it still, but the observation/action spaces are the same.
I try with and wihtout LSTM and get the same result.

I use a single agent in a MultiAgentEnv (with the intention to extend to multiagent once the single agent performs well).
The agent is a kinematic 2D puck robot with a lidar array, learning to patrol  a stationary environment. 
I use the gym interface. When reward is peaking the robot patrols the environment with good coverage (2M timesteps),
but then it quickly goes to a near non-moving robot (4M timesteps).

I understand I need to dig deeper, but maybe you have a sense or intuition what it may be?
As it works well in Stable baselines.

thank you, 
                  /Fredrik

### What is your question?

*Ray version and other system information (Python version, TensorFlow version, OS):*

**Comment by sven1977** [comment]
Hey @fredrikbissmarckatfoise , thanks for bringing this to our attention.
Would it be possible to share the two configs (baselines and RLlib) here? It could still be some default parameter having slipped through that has a different value that's causing this.

**Comment by fredrikbissmarckatfoise** [comment]
Thank you for prompt answer.

I haven't put much thought into the parameter settings as it worked well for Stable baselines, 
and initially looks good for Rllib too. It looks like it might be an overfitting problem, I might put
in some noise to see if I can regularize better.  But that implies the polices are set up rather differently 
in tensorflow.
 
Here are the parameters for each set up:

A) Stable baselines (https://github.com/hill-a/stable-baselines/blob/master/stable_baselines/ppo2/ppo2.py)

I used the default parameters:

def __init__(self, policy, env, gamma=0.99, n_steps=128, ent_coef=0.01, learning_rate=2.5e-4, vf_coef=0.5,
                 max_grad_norm=0.5, lam=0.95, nminibatches=4, noptepochs=4, cliprange=0.2, cliprange_vf=None,
                 verbose=0, tensorboard_log=None, _init_setup_model=True, policy_kwargs=None,
                 full_tensorboard_log=False, seed=None, n_cpu_tf_sess=None):
				 
I do this with MlpPolicy and MlpLstmPolicy (vectorized environment)
				 
B) RLlib (Multiagent)

This is what I specify in the application:

tune.run(
      'PPOTrainer',
      stop={'training_iteration': 100000},
      checkpoint_freq=50,
      config={
          'env': 'puckrobots',
          'lambda': 0.95,
          'kl_coeff': 0.2,
          'clip_rewards': False,
          'vf_clip_param': 1000000,
          'entropy_coeff': 0.01,
          'train_batch_size': 2000,
          'sample_batch_size': 100,
          'sgd_minibatch_size': 500,
		  'model': {'use_lstm': False} ¤ tested also for True
          'num_sgd_iter': 10,
          'num_workers': 10,
          'num_envs_per_worker': 1,
          'batch_mode': 'truncate_episodes',
          'observation_filter': 'NoFilter',
          'vf_share_layers': 'true',
          'num_gpus': 1,
          'lr': 2.5e-4,
          'log_level': 'INFO',
          'simple_optimizer': False,
          'multiagent': {
              'policies': policies,
              'policy_mapping_fn': tune.function(
                  lambda agent_id: policy_ids[int(agent_id[6:])]),
          },
      },
  )

**Comment by ericl** [comment]
Could it be learning rate annealing? Maybe stable baselines is annealing LR to zero over the time. If you do the same in RLlib the policy should also stabilize.

**Comment by fredrikbissmarckatfoise** [comment]
Thanks for taking notice of my problem. The learning rate is constant over episodes for both cases and is confirmed directly in the tensorboard display.

I now go back to the stable baselines case to make sure the test case is actually identical, and get more into the inner workings of things.

Personally, I am more interested in getting the Rllib learning to converge, than knowing the difference between the two
cases. But I think it may be of general interest what will differ between the two, if you just throw the same task on them.
Hopefully I put in the time to reach a conclusion and report it here.

/Fredrik

**Comment by fredrikbissmarckatfoise** [comment]
Looking in to it further, I found that learning is sensitive to the amplitude of the reward, in a way Stable baselines isn't. At least in the case of my particular task. Previously, I had a cumulative reward in the order of 10,000 per episode (50,000 time steps, for a somewhat successful sample). Reducing it by a factor of 1,000 (i.e. 10 per episode) and my problems with converging on the optimal behaviour are gone. That is, problem solved.

I don't know if it's by design that makes Stable baselines insensitive to the reward magnitude..

**Comment by ericl** [comment]
Is it due to value function clipping? RLlib has vf clipping enabled by default, and it is sensitive to the reward magnitude. Past the clipping threshold, the VF will learn very slowly.

**Comment by fredrikbissmarckatfoise** [comment]
Yes, vf clipping was my first suspicion. I experimented before with setting vf_clip_param to large and very large numbers, but to no avail. I also tried A2C and it had the same problem.

However, besides better understanding of the tool, it is a non-issue for me now, I can happily move forward with my Rllib work. Thanks for the concern.
